client/src/unifyfs-fixed.c

```diff
@@ -543,78 +621,77 @@ static int unifyfs_logio_chunk_write(
     /* get global file id for this file */
     int gfid = unifyfs_gfid_from_fid(fid);
 
-    /* define an new index entry for this write operation */
-    unifyfs_index_t cur_idx;
-    cur_idx.gfid     = gfid;
-    cur_idx.file_pos = pos;
-    cur_idx.log_pos  = log_offset;
-    cur_idx.length   = count;
-
-    /* lookup number of existing index entries */
-    off_t num_entries = *(unifyfs_indices.ptr_num_entries);
-
-    /* get pointer to index array */
-    unifyfs_index_t* idxs = unifyfs_indices.index_entry;
-
-    /* attempt to coalesce contiguous index entries if we
-     * have an existing index in the buffer */
-    if (num_entries > 0) {
-        /* get pointer to last element in index array */
-        unifyfs_index_t* prev_idx = &idxs[num_entries - 1];
-
-        /* attempt to coalesce current index with last index,
-         * updates fields in last index and current index
-         * accordingly */
-        unifyfs_coalesce_index(prev_idx, &cur_idx,
-            unifyfs_key_slice_range);
+    /*
+     * Store the write in our segment tree.  We will later use this for
+     * flattening writes.
+     */
+    if (unifyfs_flatten_writes) {
+        seg_tree_add(&meta->seg_tree, pos, pos + count - 1, log_offset);
     }
 
-    /* add new index entries if needed */
-    while (cur_idx.length > 0) {
-        /* remaining entries we can fit in the shared memory region */
-        off_t remaining_entries = unifyfs_max_index_entries - num_entries;
+    /* Update our write metadata with the new write */
+    return (unifyfs_logio_add_write_meta_to_index(gfid, pos, log_offset,
+        count));
+}
 
-        /* if we have filled the key/value buffer, flush it to server */
-        if (0 == remaining_entries) {
-            /* index buffer is full, flush it */
-            int ret = invoke_client_fsync_rpc(cur_idx.gfid);
-            if (ret != UNIFYFS_SUCCESS) {
-                /* something went wrong when trying to flush key/values */
-                LOGERR("failed to flush key/value index to server");
-                return UNIFYFS_ERROR_IO;
-            }
+/*
+ * Clear all entries in the log index.  This only clears the metadata,
+ * not the data itself.
+ */
+static void
+unifyfs_clear_index(void)
+{
+    size_t index_size;
+    index_size = sizeof(*unifyfs_indices.index_entry) *
+        (*unifyfs_indices.ptr_num_entries);
```

Comment:
style: let's pull (*unifyfs_indices.ptr_num_entries) into a local variable.  This helps a lot when running under a debugger like totalview.
```
    off_t num_entries = *unifyfs_indices.ptr_num_entries;
    size_t index_size = sizeof(*unifyfs_indices.index_entry) * num_entries;
```